include/universal/blas/generators/randsvd.hpp

```diff
@@ -39,27 +36,80 @@ namespace sw {
                 return rank;
             }
+            template<typename Matrix, typename MatrixQ, typename MatrixR, typename Scalar>
+            void qr(const Matrix& A, MatrixQ& Q, MatrixR& R) {
+                typedef typename matrix<Scalar>::value_type     value_type;//use Scalar
+                typedef typename matrix<Scalar>::size_type      size_type;//use size_type directly
+                typedef typename Magnitude<value_type>::type    magnitude_type;
+
+                size_type ncols = num_cols(A), nrows = num_cols(A);
+                mini = ncols == nrows ? ncols - 1 : (nrows >= ncols ? ncols : nrows);
+                magnitude_type  factor = magnitude_type(2);
+
+                Q = 1;
+                for (size_type i = 0; i < mini; ++i) {
+                    //have to compute Q && R Here
+                }
+
+            }
+
+
             template<typename Scalar>
-            void qr(const matrix<Scalar> A, matrix<Scalar> Q, matrix<Scalar> R) {
-            	//R is the upper triangular matrix
-            	//Q is the orthogonal matrix
-            	
+            std::pair<matrix<Scalar>, matrix<Scalar>>
+                inline qr(const matrix<Scalar>& A) {
+                //R is the upper triangular matrix
+                //Q is the orthogonal matrix
+                matrix<Scalar> Q(num_rows(A), num_cols(A)), R(A);
+                qr(A, Q, R);
+                return std::make_pair(Q, R);
             }
-            
             template<typename Scalar>
-            tuple randsvd(const matrix<Scalar> &A) {
+            inline void svd(const matrix<Scalar>& A, matrix<Scalar>& S, matrix<Scalar>& V, matrix<Scalar>& D, double tol = 10e-10) {
+                typedef typename matrix<Scalar>::value_type   value_type;
+                typedef typename matrix<Scalar>::size_type    size_type;
+                size_type        ncols = num_cols(A), nrows = num_rows(A);
+                value_type       ref, zero = math::zero(ref), one = math::one(ref);
+                double 	     err(std::numeric_limits<double>::max()), e, f;
+                if (nrows != ncols) std::swap(row, col);
+                matrix<Scalar> Q(nrows, nrows), R(nrows, ncols), VT(nrows, ncols), E(nrows, ncols),QT(ncols, ncols), RT(ncols, nrows);
+                size_type l = 100 * std::max(nrows, ncols);
+                S = one; D = one; E = zero;
+                for (size_type i = 0; err > tol && i < l; ++i) {
+                    std::tie(QT, RT) = qr(V);
+                    S *= QT;
+                    VT = RT.transpose();
+                    std::tie(Q, R) = qr(VT);
+                    D *= Q;
+                    E = triu(R, 1);
+                    V = trans(R);
+                    //have to implement when upper(R)=0
+                }
+            }
+            template<typename Scalar>
+            std::tuple<matrix<Scalar>, matrix<Scalar>, matrix<Scalar>>
+            inline svd(const matrix<Scalar>& A, double tol= 10e-10) {

```

Comment:
the tolerance should not be hard-typed, it should be either in the same type as the matrix elements, or if you are creating a real mixed-precision algorithm, it should be its own parameterized type.

template<typename Scalar, typename Tolerance>